openshift/system/Dockerfile

```diff
@@ -1,91 +1,88 @@
 
-ENV BASH_ENV=/opt/app-root/etc/scl_enable \
-    ENV=/opt/app-root/etc/scl_enable \
-    PROMPT_COMMAND=". /opt/app-root/etc/scl_enable" \
-    RAILS_ENV=production \
-    SAFETY_ASSURED=1
+WORKDIR /opt/system
 
-RUN source /opt/app-root/etc/scl_enable \
-    && gem install bundler --version 2.2.25 \
-    && bundle config build.pg --with-pg-config=/usr/pgsql-13/bin/pg_config \
-    && bundle config set --local without development:test:licenses \
-    && bundle install --deployment --jobs $(grep -c processor /proc/cpuinfo) --retry=5
+RUN dnf -y module enable ruby:${RUBY_VERSION} nodejs:16 mysql:8.0 \
+    && yum install -y --setopt=skip_missing_names_on_install=False,tsflags=nodocs shared-mime-info postgresql rubygem-irb rubygem-rdoc ruby libpq mysql mysql-libs gd git liberation-sans-fonts file \
+    && yum -y clean all
 
-RUN chgrp root /opt/system/
+COPY --from=builder /opt/system/ .
+COPY --from=builder /tmp/memkind-*/jemalloc/lib/libjemalloc.so* /usr/local/lib64
+COPY --from=builder /usr/local/share/gems /usr/local/share/
+COPY --from=builder /usr/local/bin/bundle* /usr/local/bin
+# in RHEL 8 podman-catatonit pulls in too many useless deps so we don't install the RPM directly
+COPY --from=builder /usr/libexec/catatonit/catatonit /usr/libexec/catatonit/catatonit
 
-ADD . ./
-ADD package.json ./
-ADD yarn.lock ./
-ADD openshift/system/config/* ./config/
+RUN echo /usr/local/lib64 > /etc/ld.so.conf.d/jemalloc.conf \
+    && ldconfig && ldconfig -p | grep jemalloc
 
-RUN source /opt/app-root/etc/scl_enable \
+RUN bundle config set --local without development:test:licenses:assets \
     && bundle exec rake tmp:create \
-    && mkdir -p public/assets db/sphinx \
+    && mkdir -p db/sphinx \
     && chmod g+w -vfR log tmp public/assets db/sphinx \
     && umask 0002 \
-    && cd /opt/system \
-    && npm install -g yarn \
-    && yarn install:safe --no-progress \
-    && SECRET_KEY_BASE=rails/32947 bundle exec rake assets:precompile tmp:clear \
-    && rm log/*.log \
+    && rm -f log/*.log \
     && chmod g+w /opt/system/config
 
-ADD openshift/system/entrypoint.sh /opt/system/entrypoint.sh
+# Bundler runs git commands on git dependencies
+# https://bundler.io/guides/git.html#local-git-repos
+# git will check if the current user is the owner of the git repository folder
+# This was included in git v2.35.2 or newer.
+# https://github.com/git/git/commit/8959555cee7ec045958f9b6dd62e541affb7e7d9
+# Openshift changes the effective UID, so this git check needs to be bypassed.
+RUN git config --global --add safe.directory '*'
+
+# container-entrypoint is expected by the operator so keeping it for compatibility
+RUN ln -s openshift/system/entrypoint.sh . \
+    && ln -s /opt/system/openshift/system/contrib/container-entrypoint /usr/local/bin/
+
 USER 1001
 EXPOSE 3000 9306
-# TODO: dumb-init!
-ENTRYPOINT ["/opt/system/entrypoint.sh"]
+ENTRYPOINT ["/usr/libexec/catatonit/catatonit", "--", "/opt/system/entrypoint.sh"]
```

Comment:
What's this `catatonit` thing about?

Reply:
You can check catatonit documentation but probably what you ask will be better answered by
https://github.com/Yelp/dumb-init#why-you-need-an-init-system

Reply:
Thanks for the link. I'm surprised that kind of problems are not solved by the container engine itself.

Reply:
While signal handling might be fixable, I assume the zombie thing is more complicated because they can't know whether process 1 will care about dead children or not.

Reply:
Yeah but couldn't the engine include a fake init as PID 1 and launch the command as PID 2?

Reply:
You never know if the user process would want to keep track of indirect unhandled children or not. Would be rare but a possible use case. Actually Podman inserts `catatonit` when `--init` option is provided on the command line. But not automatically. Would have been nice to have kubernetes also do so when container has some option specified.

Reply:
One example would be a container with a standard init system (e.g `systemd`) inside.